dev/fattest.simplicity/src/componenttest/security/utils/SSLUtils.java

```diff
@@ -0,0 +1,175 @@
+        final String IPV4_PATTERN = "^(([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])(\\.(?!$)|$)){4}$";
+
+        Provider bcProvider = new BouncyCastleProvider();
+        Security.addProvider(bcProvider);
+
+        long now = System.currentTimeMillis();
+        Date startDate = new Date(now);
+
+        X500Name dnName = new X500Name(subjectDN);
+
+        // Using the current timestamp as the certificate serial number
+        BigInteger certSerialNumber = new BigInteger(Long.toString(now));
+
+        Calendar calendar = Calendar.getInstance();
+        calendar.setTime(startDate);
+        // 1 Yr validity
+        calendar.add(Calendar.YEAR, 1);
+
+        Date endDate = calendar.getTime();
+
+        // Use appropriate signature algorithm based on your keyPair algorithm.
+        String signatureAlgorithm = "SHA256WithRSA";
+
+        SubjectPublicKeyInfo subjectPublicKeyInfo = SubjectPublicKeyInfo.getInstance(keyPair
+                .getPublic().getEncoded());
+
+        X509v3CertificateBuilder certificateBuilder = new X509v3CertificateBuilder(dnName,
+                certSerialNumber, startDate, endDate, dnName, subjectPublicKeyInfo);
+
+        List<GeneralName> altNames = new ArrayList<GeneralName>();
+        for (String name : genericNameList) {
+            if (name.matches(DNS_PATTERN)) {
+                Log.info(c, "selfSign", "adding dns GeneralName " + name);
+                altNames.add(new GeneralName(GeneralName.dNSName, name));
+            } else if (name.matches(IPV4_PATTERN)) {
+                Log.info(c, "selfSign", "adding ip GeneralName " + name);
+                altNames.add(new GeneralName(GeneralName.dNSName, name));
+            }
+        }
+
+        GeneralNames subjectAltNames = GeneralNames.getInstance(new DERSequence((GeneralName[]) altNames.toArray(new GeneralName[] {})));
+        certificateBuilder.addExtension(Extension.subjectAlternativeName, false, subjectAltNames);
+
+        ContentSigner contentSigner = new JcaContentSignerBuilder(signatureAlgorithm).setProvider(
+                bcProvider).build(keyPair.getPrivate());
+
+        X509CertificateHolder certificateHolder = certificateBuilder.build(contentSigner);
+
+        Certificate selfSignedCert = new JcaX509CertificateConverter()
+                .getCertificate(certificateHolder);
+
+        return selfSignedCert;
+    }
+
+    public static void exportPublicKeyToFile(String path, KeyPair keyPair) throws Exception {
+        Log.info(c, "exportPublicKeyToFile", "entering SSLUtils.exportPublicKeyToFile");
+
+        PublicKey publicKey = keyPair.getPublic();
+        File file = new File(path);
+        file.getParentFile().mkdirs();
+        file.createNewFile();
+
+        FileWriter fw = new FileWriter(file, false);
+        JcaPEMWriter pemWriter = new JcaPEMWriter(fw);
+        pemWriter.writeObject(publicKey);
```

Comment:
You're not closing the stream.

You _must_ always ensure that any kind of IO stream is closed. You should use try-with-resources where possible.

```suggestion
        try (FileWriter fw = new FileWriter(file, false);
             JcaPEMWriter pemWriter = new JcaPEMWriter(fw)) {
            pemWriter.writeObject(publicKey);
        }
```

Reply:
That will actually cause an error. PemWriter extends BufferedWriter and BufferedWriter.close() calls Writer.close() so we'd get an IOException from closing BufferedWriter twice. 

But using try-with-resources for the PemWriter would be a good idea.

Reply:
Calling `Closable.close()` has no effect if the stream is already closed so it shouldn't cause an error.

Only closing the `JcaPEMWriter` should be fine though as long as the doc says that it will close the underlying writer.